### ISSUE TITLE
Table Pagination slow on sort

### ISSUE THREAD

**Issue by ndean2017** [bootstrap-vue/bootstrap-vue#2990]
When sorting a column in a b-table with pagination, the sort does not actually change the order of the b-table items, meaning that every time you change the page 1->2->3 it is resorting each time. This isn't apparent unless there is a large number of rows. I used about ~15,000 rows in my case but it would be very clear in a case with 100k or more.

**Comment by tmorehouse** [comment]
If the underlying items array hasn't changed, then the sorting computed prop should have the results cached. But if you are using a provider function (rather than an array of items), then sorting may occur more often.

For large arrays of items, it is recommended to use server side sorting and pagination.

**Comment by jackmu95** [comment]
Here is a basic demo with 15000 items and no issue:
https://codesandbox.io/s/ryo392x46q

**Comment by tmorehouse** [comment]
Here is the order of the computed props:

```
Original Items Array => Filtering => Sorting => Pagination
```

So any changes to props that would trigger a change in the steps above would cause sorting and pagination to occur again.

**Comment by ndean2017** [comment]
I haven't been able to get a sandbox working properly yet with vue-property-decorator, typescript, and bootstrap but here is a code snippet of what I am doing. Up the chain a little, items is populated by a computed function.

`<template>
  <div class="symbols overflow-auto">
    <b-row>
      <b-col>
        <b-table small outlined
                 id="symbolsTable"
                 :items="items"
                 :fields="fields"
                 :current-page="currentPage"
                 :per-page="perPage"
        >
        </b-table>
        <b-pagination class="floatRight"
                      v-model="currentPage"
                      :total-rows="numRows"
                      :per-page="perPage"
                      aria-controls="symbolsTable">
        </b-pagination>    
      </b-col>
    </b-row>
  </div>
</template>

<script lang="ts">

import { Vue, Component, Prop } from 'vue-property-decorator'

@Component({
  components: { }
})
export default class Symbols extends Vue {
  @Prop() items! : object[] // table data
  fields: any = [
    { key: 'native_symbol', label: "Native Symbol", sortable: true},
    { key: 'symbol_1', label: "Symbol 1", sortable: false },
    { key: 'symbol_2', label: "Symbol 2", sortable: false },
    { key: 'symbol_3', label: "Symbol 3", sortable: false },
    { key: 'symbol_4', label: "Symbol 4", sortable: false },
    { key: 'service_level', label: "Service Level"}
  ]
  currentPage: number = 1
  perPage: number = 20
  numRows: number = this.items.length
}

</script>`

To try and isolate the issue I made a deep clone of the data to another variable before passing it in as a property: `this.data = _.cloneDeep(this.db.symbols)`

**Comment by tmorehouse** [comment]
Is the cloning a computed property or a method?

If it is a method, anytime something updates your component (props, data changes, re-render required, etc), that will cause the method to be re-called, creating a new cloned array instance reference, which will in turn trigger the filtering, sorting and pagination sequence to occur again.

**Comment by tmorehouse** [comment]
@ndean2017 how did you determine it was re-sorting the data each time the page number changed?

**Comment by tmorehouse** [comment]
the pagination routine uses the native browser `Array.prototype.slice()` method to select hte page of data to display.

Different browser types (and versions) have different performance for `array.slice()` based on their implementation.  You might be running into a browser specific performance issue.

**Comment by tmorehouse** [comment]
https://jsperf.com/converting-arguments-to-an-array/30

**Comment by ndean2017** [comment]
Once I get the rows from my database I would normally populate the prop directly from a computed property. I switched this to deepClone which is purely a method that will only ever be called once when the before the component is created. I determined it was sorting every time by clicking the arrows on one of the columns to initially sort the rows, followed by starting a performance recording in the chrome debugger which I attached below. The recording was started after the sort prior to changing from page 1-2
![Capture](https://user-images.githubusercontent.com/49209916/55484531-a29c9f80-55f6-11e9-82d3-f8dffb119768.PNG)
